File: SPRING/Hospital_Management_system_angular/src/app/services/lab-config.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import { InterpretationRule, LabRule, ReferenceRange, TestMasterDetail, TestParameter } from '../models/lab-config.model';

@Injectable({ providedIn: 'root' })
export class LabConfigService {

  private api = environment.apiUrl + 'lab/config';

  constructor(private http: HttpClient) {}

  getTestMasterDetail(testMasterId: number): Observable<TestMasterDetail> {
    return this.http.get<TestMasterDetail>(`${this.api}/test-masters/${testMasterId}`);
  }

  getAllParameters(): Observable<TestParameter[]> {
    return this.http.get<TestParameter[]>(`${this.api}/parameters`);
  }

  getParameter(id: number): Observable<TestParameter> {
    return this.http.get<TestParameter>(`${this.api}/parameters/${id}`);
  }

  createParameter(data: TestParameter): Observable<TestParameter> {
    return this.http.post<TestParameter>(`${this.api}/parameters`, data);
  }

  updateParameter(id: number, data: TestParameter): Observable<TestParameter> {
    return this.http.put<TestParameter>(`${this.api}/parameters/${id}`, data);
  }

  deleteParameter(id: number): Observable<any> {
    return this.http.delete(`${this.api}/parameters/${id}`);
  }

  createRange(data: ReferenceRange): Observable<ReferenceRange> {
    return this.http.post<ReferenceRange>(`${this.api}/reference-ranges`, data);
  }

  updateRange(id: number, data: ReferenceRange): Observable<ReferenceRange> {
    return this.http.put<ReferenceRange>(`${this.api}/reference-ranges/${id}`, data);
  }

  deleteRange(id: number): Observable<any> {
    return this.http.delete(`${this.api}/reference-ranges/${id}`);
  }

  createRule(data: InterpretationRule): Observable<InterpretationRule> {
    return this.http.post<InterpretationRule>(`${this.api}/interpretation-rules`, data);
  }

  updateRule(id: number, data: InterpretationRule): Observable<InterpretationRule> {
    return this.http.put<InterpretationRule>(`${this.api}/interpretation-rules/${id}`, data);
  }

  deleteRule(id: number): Observable<any> {
    return this.http.delete(`${this.api}/interpretation-rules/${id}`);
  }

  getLabRules(): Observable<LabRule[]> {
    return this.http.get<LabRule[]>(`${this.api}/rules`);
  }

  createLabRule(data: LabRule): Observable<LabRule> {
    return this.http.post<LabRule>(`${this.api}/rules`, data);
  }

  updateLabRule(id: number, data: LabRule): Observable<LabRule> {
    return this.http.put<LabRule>(`${this.api}/rules/${id}`, data);
  }

  deleteLabRule(id: number): Observable<any> {
    return this.http.delete(`${this.api}/rules/${id}`);
  }
}

File: SPRING/Modern_Hospital_Management_System/src/main/java/emranhss/com/Modern_Hospital_Management_System/service/lab/LabConfigService.java
package emranhss.com.Modern_Hospital_Management_System.service.lab;

import emranhss.com.Modern_Hospital_Management_System.dto.request.InterpretationRuleRequest;
import emranhss.com.Modern_Hospital_Management_System.dto.request.LabRuleRequest;
import emranhss.com.Modern_Hospital_Management_System.dto.request.ReferenceRangeRequest;
import emranhss.com.Modern_Hospital_Management_System.dto.request.TestParameterRequest;
import emranhss.com.Modern_Hospital_Management_System.dto.response.*;

import java.util.List;

public interface LabConfigService {

    TestMasterDetailResponse getTestMasterDetail(Long testMasterId);

    List<TestParameterResponse> getAllParameters();

    TestParameterResponse getParameter(Long id);

    TestParameterResponse createParameter(TestParameterRequest request);

    TestParameterResponse updateParameter(Long id, TestParameterRequest request);

    void deleteParameter(Long id);

    ReferenceRangeResponse createRange(ReferenceRangeRequest request);

    ReferenceRangeResponse updateRange(Long id, ReferenceRangeRequest request);

    void deleteRange(Long id);

    InterpretationRuleResponse createRule(InterpretationRuleRequest request);

    InterpretationRuleResponse updateRule(Long id, InterpretationRuleRequest request);

    void deleteRule(Long id);

    List<LabRuleResponse> getAllLabRules();

    LabRuleResponse createLabRule(LabRuleRequest request);

    LabRuleResponse updateLabRule(Long id, LabRuleRequest request);

    void deleteLabRule(Long id);
}
